content/ADR-203-social-ws-login.md

```diff
@@ -0,0 +1,158 @@
+
+This is an exhaustive analysis for the Authentication section of [Social Service M2](/adr/ADR-189).
+
+## Context, Reach & Prioritization
+
+Authentication will be done using the same mechanism as Matrix, meaning that the client is responsible for obtaining a token that must be sent to the service, leveraging the work in [Social Service Authentication](/adr/ADR-143).
+
+Once the token is received, the service queries Synapse to retrieve the corresponding `user_id` and handles all queries using that user's credentials. To prevent overloading Synapse, this information is cached in Redis.
+
+The document exposes possible solutions for sending the token from the Client to the Server.
+
+## Solution Space Exploration
+
+### Solution 1.a: use a login message
+
+To send the token, this solution proposes to expand the current `.proto` definition with a mandatory login message that must be sent as the first message when establishing a WebSocket connection. The client sends its token in this message, and the service validates it against Synapse to obtain the corresponding `user_id`. This `user_id` remains the same throughout the connection for all subsequent messages, taking advantage of the established WebSocket connection.
+
+Example of proto file:
+
+```proto
+service FriendshipsService {
+  rpc Login(Token) returns (google.protobuf.Empty) {}
+  rpc GetFriends(google.protobuf.Empty) returns (stream Users) {}
+}
+```
+
+#### Solution 1: Advantages
+
+- Takes advantage of the existing state and connection (establishes a relation one to one from connection to user, anyway this have some drawbacks explained in the disadvantages section).
+- Saves sending the token in the payload for each message (only a few bytes), avoiding the need to fetch it from Redis.
+- If a new connection with the same token/user_id arrives, the previous connection can be closed. This is a simple way to avoid DoS.
+
+#### Solution 1: Disadvantages
+
+- The information is stored in memory, which may be a replication of what is in Redis.
+- [dcl-rpc](https://github.com/decentraland/rpc-rust) needs to be expanded to allow connection identification and specific information storage. Currently, a global context is stored for the entire service, but a way to identify the client (there is no IP address or socket, only user_id) needs to be found. Additionally, `dcl-rpc` currently does not expose the connection because it is abstracted.
+- Client-side logic is more complex, as it must ensure that the first message sent is the login message.
+- As the service has a state, then it makes more complex to scale horizontally, as a new login is required when reconnecting with another node, even using the same token.
+- Websockets support automatic reconnection, in this case if the server restarts and loses the memory state and reconnects to the user, then the login information will be lost.
```

Comment:
```suggestion
- The information is stored in memory, which may be a replication of what is in Redis.
- [dcl-rpc](https://github.com/decentraland/rpc-rust) needs to be expanded to allow connection identification and specific information storage. Currently, a global context is stored for the entire service, but a way to identify the client (there is no IP address or socket, only user_id) needs to be found. Additionally, `dcl-rpc` currently does not expose the connection because it is abstracted.
- Client-side logic is more complex, as it must ensure that the first message sent is the login message.
- As the service has a state, then it makes it more complex to scale horizontally, as a new login is required when reconnecting with another node, even using the same token.
- Websockets support automatic reconnection, in this case, if the server restarts and loses the memory state and reconnects to the user, then the login information will be lost.
```